osmet: efficient packing/unpacking of CoreOS metal images

jlebon opened the pull request:
Introduce a new `osmet` command with two primary subcommands: `pack` and
`unpack`. Together, these allow offline bare metal installation of
CoreOS images while only slightly increasing the size of the initramfs.

The `pack` subcommand takes as input a block device of a CoreOS metal
image and the expected checksum to match. It mounts the root partition
from that device and generates a smaller version of the metal image
itself with the OSTree objects "punched out" (this is called the
"punched image" in the code). The command outputs this smaller version
as well as a lookup table of where the OSTree objects belonged into an
"osmet" binary file.

The `unpack` subcommand takes as input an osmet binary file and a path
to an OSTree repo and reconstrust the metal image, bit for bit. This
command is more for testing in practice. The following patch will teach
the `install` command to use the osmet path by default, which is how
users will interact with this.

jlebon commented:
This is also missing scanning the `/boot` partition, which should take the size of the osmet binary further down from 80M to 10M. But the basic concept works otherwise!

jlebon commented:
See the cosa side of this at https://github.com/coreos/coreos-assembler/pull/1244. That's one way of running this. Though you can also run the `osmet-pack` script there directly, e.g.:

```
$ cosa buildextend-metal
$ /usr/lib/coreos-assembler/osmet-pack builds/latest/x86_64/*-metal*.x86_64.raw tmp/osmet.bin <checksum>
...
Total OSTree objects scanned from root: 14614 (14609 mapped, 5 empty)
Total extents: 14609
Duplicate extents dropped: 0
Overlapping extents clamped: 0
Total bytes skipped: 1621889024
Total bytes written: 1237577728
Total bytes written (compressed): 82599292
$ ls -lh tmp/osmet.bin
-rw-------. 1 jlebon jlebon 80M Mar 13 17:01 tmp/osmet.bin
```

Or you can just bring up a VM with the metal image attached:

```
cosa run -- -drive if=none,format=raw,file=path/to/metal.raw,id=foo -device virtio-blk,drive=foo
```

And play with the `coreos-installer osmet` in there.

lucab commented on src/cmdline.rs:
```diff
@@ -341,6 +427,12 @@ pub fn parse_args() -> Result<Config> {
             ("remove", Some(matches)) => parse_iso_remove(&matches),
             _ => bail!("unrecognized 'iso' subcommand"),
         },
+        ("osmet", Some(osmet_matches)) => match osmet_matches.subcommand() {
```
Longer term, how do you plan to arrange the CLI? My understanding is that here in `coreos-installer` we will want to expose only the porcelain entrypoint on top of the current `unpack` logic, right? If so, where does the mapping/packing belong?

jlebon replied:
See the last paragraph of https://github.com/coreos/coreos-installer/pull/187#issue-387995035. We could mark the `osmet` command as hidden in the end.

Packing happens in cosa: https://github.com/coreos/coreos-assembler/pull/1244.

lucab replied:
Ack.
I think I was more doubtful with regards to the packing logic. That sounds like something which should be implemented in cosa/mantle codebase (i.e. in Go, this touches into the `bincode` discussion below).
Or do you see any usecase where you need the packing logic on a FCOS host?

jlebon replied:
The reason for having the packing logic here too is because it makes sense for it to be colocated with the code that will read it in. There's lots of code sharing going on here. (Notice for example how `osmet_pack` essentially does a full unpacking at the end as a sanity check; that uses the same code as `osmet_unpack` would).

A nice advantage of this for example is that it naturally binds versioning of the osmet file with the osmet code paths since cosa itself will use the baked coreos-installer in the image to do the packing.

lucab commented on src/osmet.rs:
```diff
@@ -0,0 +1,809 @@
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct OsmetFile {
```
How do we define/handle endianess on all of this?

jlebon replied:
`bincode` automatically canonicalizes to little endian. I can create a [`Config`](https://docs.rs/bincode/1.2.1/bincode/struct.Config.html) object to make that more explicit (and in case the default changes, though that would be a breaking change).

lucab replied:
It sounds like a good idea. Or at least mention that somewhere in docs/comments.

lucab commented on src/osmet.rs:
```diff
@@ -0,0 +1,809 @@
+        osmet,
+    };
+
+    bincode::serialize_into(&mut f, &osmet_file).chain_err(|| "failed to serialize osmet file")?;
```
I'm a bit worried about using `bincode` here for data at rest, as I fear it has no interop story outside of Rust. Can we pick somebody a bit more common?

jlebon replied:
Hmm, do you have any recommendations? I really like `bincode`'s dead-simple approach and am not sure if interop outside of Rust should be a concern given how specific to this code the binary format is. (It's not like a generic JSON file which would be useful on its own.)

lucab replied:
I somehow see this as an interop wrapper between the containerized producing side (cosa/mantle) and the on-host consuming side (coreos-installer). Clearly not the same concerns as public metadata, but still something that benefits from a stable and easy format for both sides. I had protobuf in mind, even if I'm not a great fan of it.

jlebon replied:
Do the discussions in https://github.com/coreos/coreos-installer/pull/187#discussion_r392888346 clear this up?

lucab replied:
The comment at https://github.com/coreos/coreos-installer/pull/187#issuecomment-607445367 did. We are taking all precautions so that (so far) this is not in any way a public API.

jlebon commented:
I haven't addressed @lucab's review comments yet, but I fixed up all the `XXX` markers. We now also scan `/boot` for OSTree objects; this brings the total size of the osmet binary to 5.4M for FCOS!

jlebon commented:
Just noting here to remember for myself: in a conversation with @lucab, he mentioned it'd be useful to have a `doc/` entry for this, similarly to the ISO Ignition config embedding functionality. I agree.

jlebon commented on src/blockdev.rs:
```diff
@@ -30,28 +32,40 @@ use tempfile;
 
 use crate::errors::*;
 
-pub fn mount_boot(device: &str) -> Result<Mount> {
```
Yeah, definitely a couple of prep patches I need to split out.

jlebon replied:
OK, prep patches in https://github.com/coreos/coreos-installer/pull/191.

jlebon commented on src/blockdev.rs:
```diff
@@ -153,6 +167,42 @@ impl Mount {
+}
+
+fn read_sysfs_dev_block_value(maj: u64, min: u64, field: &str) -> Result<String> {
+    let path = PathBuf::from(format!("/sys/dev/block/{}:{}/{}", maj, min, field));
```
I will sanity-check, but it should since privileged containers normally bring in sysfs (at least on my Silverblue).

jlebon replied:
Yup, works!

jlebon commented on src/fiemap.rs:
```diff
@@ -0,0 +1,159 @@
+
+/// Returns the `Extent`s associated with the given file. Note that the physical offsets are
+/// relative to the partition start on which the file resides.
+pub fn fiemap(fd: RawFd) -> Result<Vec<Extent>> {
```
Thanks! Weird I somehow missed that. Will take a look and compare approaches.

jlebon replied:
OK, had a look at this. Some notes:

- Uses an iterator interface, which is neat. (One minor downside is that it hurts readability a bit because state maintenance is scattered across functions rather than just in one big loop).
- Uses fewer of the FFI facilities available. For example, it directly declares the `ioctl` function as an external function instead of the `nix::sys::ioctl` helpers and encodes the `FS_IOC_FIEMAP` constant instead of deriving it via the `nix` macros.

Overall, I don't find it obviously better (but TBF, not obviously worse either). We could probably upstream tweaks from here and rebase on top of that. Though not sure if I want to block on that.

jlebon commented on src/blockdev.rs:
```diff
@@ -139,7 +153,7 @@ impl Mount {
             &mountpoint,
             Some(fstype),
-            mount::MsFlags::empty(),
+            flags.unwrap_or_else(mount::MsFlags::empty),
```
I addressed this in https://github.com/coreos/coreos-installer/pull/191.

cgwalters commented on src/fiemap.rs:
```diff
@@ -0,0 +1,159 @@
+
+    impl fiemap {
+        pub fn new() -> Self {
+            let mut r: Self = unsafe { zeroed() };
```
Or just use `Default::default()`.

jlebon replied:
It's not so much to save a few lines of code. It's just awkward to initialize a large array of elements without any copying. Now at least there is `MaybeUninit`, see e.g.: https://doc.rust-lang.org/core/mem/union.MaybeUninit.html#initializing-an-array-element-by-element. But note even that is unsafe. I can switch to that model though if preferred!

cgwalters commented on src/osmet.rs:
```diff
@@ -0,0 +1,969 @@
+}
+
+// The sha2 crate wants us to use generic-array and typenum to deal with checksums, which are ways
+// to hack around Rust's lack of const generics. We sidestep that goop here in favour of the simple
```
To optimize compliation time/number of dependencies?

jlebon replied:
Mostly complexity. I initially did use generic arrays, though it felt kinda painful to work with.

cgwalters commented on src/osmet.rs:
```diff
@@ -0,0 +1,969 @@
+use error_chain::bail;
+use nix::mount;
+use serde::{Deserialize, Serialize};
+use sha2::{Digest, Sha256};
```
I think we should at least consider a build-time option to use openssl.  I doubt the `sha2` crate is FIPS certifed for example.

jlebon replied:
Hmm, hadn't considered this. Yes, will investigate. (One thing I noticed is that `sha2` takes much longer to hash than `sha256sum`. Maybe using openssl would help with this.)

cgwalters replied:
(Depending on openssl on the other hand makes the binary much less portable)

jlebon replied:
Hmm, though do we support running coreos-installer as a fat binary? I think the only supported paths are in the live ISO/PXE or the container (or build it yourself of course). (/cc @bgilbert) E.g. we already link to liblzma right now for example.

jlebon replied:
OK nice, not surprisingly OpenSSL is at least as fast as `sha256sum`.

cgwalters commented on src/osmet.rs:
```diff
@@ -0,0 +1,969 @@
+
+pub fn osmet_pack(config: &OsmetPackConfig) -> Result<()> {
+    // First, mount the two main partitions we want to suck out data from: / and /boot. Note
+    // MS_RDONLY; this also ensures that the partition isn't already mounted rw elsewhere.
```
Best practice here would be for the invoking program to use `unshare -m` right?

jlebon replied:
Yes, agreed. Hmm, and I guess since this has to run as root, we could fold the unshare logic here. Though will leave that as a follow up.

jlebon replied:
Split that out into https://github.com/coreos/coreos-installer/issues/208.

cgwalters commented on src/osmet.rs:
```diff
@@ -0,0 +1,969 @@
+    let mut mappings: Vec<Mapping> = vec![];
+    let mut mapped_file_count = 0;
+    let mut empty_file_count = 0;
+    for entry in WalkDir::new(objects_dir) {
```
I'm fine with this but... I think it's also poking into ostree internals.  It's not like we are likely to change the layout (doing so would be *really hard* due to backcompat constraints) but...consider using  https://crates.io/crates/ostree maybe?

jlebon replied:
I considered this a bit, but `ostree_repo_list_objects` is basically doing the same thing except it's incurring way more overhead from handling all object types and allocating lots of `GVariant`s.

The other thing is that, we don't really care about what commits are in the repo for example. Whatever file object we find is clearly on the disk and so would benefit from mapping. That said, querying libostree for commit information could still be used before unpacking as a sanity-check that we'll have all the objects we need. I can definitely see linking to libostree in this code in the future though. Just not sure if we need it just yet.

cgwalters commented on src/osmet.rs:
```diff
@@ -0,0 +1,969 @@
+fn prescan_boot_partition(boot: &Mount) -> Result<HashMap<u64, PathBuf>> {
+    let mut files: HashMap<u64, PathBuf> = HashMap::new();
+
+    for entry in WalkDir::new(boot.mountpoint()) {
```
Personally I love using rayon for this type of stuff Just Because We Can and it's easy.

jlebon replied:
Yeah, could do this. Though... this is so far from the hot path compared to other parts, not sure if it's worth the added complexity, even if the API is easy.

jlebon commented on src/cmdline.rs:
```diff
@@ -473,3 +565,46 @@ fn parse_iso_remove(matches: &ArgMatches) -> Result<Config> {
+        file: matches
+            .value_of("file")
+            .map(String::from)
+            .expect("file missing"),
```
All of these arguments are marked as required, so clap already gives a nice error message if it's missing. This matches the rest of the code in that file.

jlebon commented:
OK, marking this as ready for review! I've had to rework unpacking to make it happen in a thread so that we can just reuse the same `ImageSource` on the `install.rs` side as other sources. Another notable change is that I've made the osmet code its own Rust submodule and split it across three files to make it easier to find one's way around.

The second commit here is a key change: we teach `coreos-installer install` about using osmet as an image source, and make it the default installation method if it's available.

jlebon commented:
Some follow-ups to this:
- make sure this works with metal4k (the groundwork is there to do this, but I haven't tested it yet)
- possibly catch the optimization for when installing by stream is requested and it already matches our version (and so we can use the embedded osmet file)
- make sure we're testing this in this upstream CI (though this repo is already doing `kola testiso`, so we should get it for free once we [adapt that](https://github.com/coreos/coreos-assembler/pull/1244#issuecomment-604550082))

jlebon commented:
Hmm actually, I think we'll need to switch to it as the default as a separate step to help ratchet this in. So the order would be:
- get `osmet pack` in (first commit here) and respin coreos-installer in FCOS
- get the cosa patch in that runs `osmet pack` and bakes in the osmet files
- get CI to test this using some env var
- at some later time, once we're confident it's baked enough, we can drop the env var and make it the default

jlebon commented:
OK, dropped that commit!

jlebon commented:
CI fix in https://github.com/coreos/coreos-installer/pull/198.
And fixed the unused warnings due to dropping the second commit!

jamescassell commented:
How is it different than jigdo?

cgwalters commented:
> How is it different than jigdo?

It's a more targeted and efficient solution for what we need to start.

See also [rojig](https://github.com/coreos/rpm-ostree/issues/1081) BTW which is inspired by jigdo, but (a lot like osmet) is optimized around OSTree and not debs/rpms.

cgwalters commented:
Can you confirm that we're not truly committing to any kind of ABI around osmet?  In the end we will ship a `coreos-installer` binary and a `.osmet` file as part of the live rootfs that were versioned/built together?  (Which I guess gets into coreos-assembler using the *target* `coreos-installer` and not the one in the container image)

And another question - we're still verifying the checksum of the resulting image, right?  So at worst, osmet could fail to work, but would not write a corrupted image.

jlebon commented:
> Can you confirm that we're not truly committing to any kind of ABI around osmet?

Yes, the osmet file isn't and shouldn't for the time being be a distinct public artifact. Importantly, we can assume a strong binding between the packing and unpacking paths (which as you mentioned is why https://github.com/coreos/coreos-assembler/pull/1244 usies the coreos-installer binary from the tree itself). The versioning in the file format is more as a good practice than anything.

Concretely then, this means that we are free to change the implementation details of how offline installs work. Essentially the only API that we're making public with this is that there *is* an offline path that uses just the live ISO as the artifact.

> And another question - we're still verifying the checksum of the resulting image, right? So at worst, osmet could fail to work, but would not write a corrupted image.

Correct. More precisely, the following checks are in place:
1. `cosa buildextend-live` passes the expected checksum of the block device, which matches the checksum of the raw file itself from `meta.json`
2. after packing, `osmet pack` does a final check to verify that re-unpacking the data yields the same checksum
3. the checksum is written to the osmet header
4. `osmet unpack` verifies during unpacking that the checksum from the header matches the final checksum of all the bytes written to the provided writer. This keys into the regular `coreos-installer install` behaviour of keeping the first 1M unwritten in case the checksums don't match.

lucab commented on src/cmdline.rs:
```diff
@@ -329,6 +348,82 @@ pub fn parse_args() -> Result<Config> {
+                        )
+                        .arg(
+                            // XXX: rebase on top of
+                            // https://github.com/coreos/coreos-installer/pull/178 and use the same
```
That PR takes SHA512 only (like Ignition), though. Should we switch this PR to align on that?

jlebon replied:
I mostly meant using the same `type-digest` format and possibly factoring out the `IgnitionHash` bit and its `try_parse` there to handle SHA256. In practice, cosa stores SHA256 digests for artifacts so that's likely what'll get passed in so we don't have to recalculate a different digest. And OSTree filepath objects are also SHA256. But at least encoding `sha256` in the argument string will make it that bit easier to move to `sha512` or whatever else one day.

jlebon replied:
#178 is merged, though let's address this in a follow-up to try to get this in sooner. Going to mark this as resolved for now to help track progress here.

bgilbert commented on src/osmet/mod.rs:
```diff
@@ -0,0 +1,579 @@
+
+    assert!(cursor <= partition.end_offset);
+
+    // and now just transfer the rest of the the partition and disk
```
`and disk` is false, correct?

`the the`

bgilbert commented on src/osmet/mod.rs:
```diff
@@ -0,0 +1,579 @@
+}
+
+fn canonicalize(mappings: &mut Vec<Mapping>) {
+    assert!(!mappings.is_empty());
```
Nothing in the callers prevent this.  Should we just return early?

bgilbert commented on src/osmet/mod.rs:
```diff
@@ -0,0 +1,579 @@
+        let mut buf = [0u8; 50];
+
+        let data = [0u8; 30];
+        copy_exactly_n(&mut &data[..], &mut sink, 0, &mut buf).unwrap();
```
Wouldn't hurt to assert return values for all of these.

bgilbert commented on src/osmet/unpacker.rs:
```diff
@@ -0,0 +1,254 @@
+        Self::new_impl(osmet, punched_image, repo)
+    }
+
+    fn new_impl(
```
Ah, I see it's used in #197.

bgilbert commented on src/osmet/mod.rs:
```diff
@@ -0,0 +1,579 @@
+    /// Size of the final disk image. Used for unpacking verification.
+    size: u64,
+    /// Required sector size of target block device during unpacking.
+    sector_size: u32,
```
We also have a dependency on the CPU architecture of the target, so we should record that as well.  It's possible for `coreos-installer` to write an image with a different arch than its own, e.g. when writing to a USB stick.

jlebon replied:
I think it's worth clarifying that technically neither packing nor unpacking have any dependency on the intended CPU architecture of the source disk image. I.e. it's entirely feasible to run `coreos-installer osmet pack` on aarch64 while pointing it at a mounted x86_64 disk.

That said, I agree with you we should pick up that info as a sanity-check on unpacking. The shortcut I took for now is to just assume that the node which runs `coreos-installer osmet pack` is on the same architecture as the one intended for the disk we're packing. This is true today because cosa doesn't support cross-building.

Anyway, with that, `coreos-installer install --architecture <some other arch>` won't use osmet since it doesn't match.

I also added a comment about using `coreos-assembler.basearch` from the base commit metadata would be more correct (or we could just add `basearch` to the aleph JSON file too).

bgilbert commented on src/osmet/mod.rs:
```diff
@@ -0,0 +1,579 @@
+#[derive(Serialize, Deserialize, Debug)]
+struct Osmet {
+    /// Partitions for which we've registered mappings.
+    partitions: Vec<OsmetPartition>,
```
At the moment, `find_matching_osmet_in_dir` needs to load all the mappings before it can check the header for the desired values.  Would it make sense to restructure the osmet file as a header plus a separately-serialized body?

jlebon replied:
Yeah, good call. Goes well with the other header rework mentioned earlier.

bgilbert commented on src/osmet/mod.rs:
```diff
@@ -0,0 +1,579 @@
+    size: u64,
+    /// Required sector size of target block device during unpacking.
+    sector_size: u32,
+}
```
Would it make sense to include the contents of `/etc/os-release`?

jlebon replied:
Can you expand on this? More for making inspecting an osmet file by hand more palatable, or for some kind of sanity-checking?

bgilbert replied:
I was originally thinking of an osmet file as a longer-lived artifact, and thought we might want it to be more self-describing.  Since that isn't the goal, I don't think it's necessary to have a machine-readable summary of the image.

However, there's still https://github.com/coreos/coreos-installer/pull/197/files#r402686439.  One approach could be to add a `description` header field, containing a string like `Fedora CoreOS 31.20200323.2.1 x86_64, 512-byte sectors`.  It'd be cleaner to have `osmet pack` take that from a command-line arg, rather than parsing `os-release` itself.

Or we could just encode all those details in the osmet filename and print that from `Display`.

bgilbert commented on src/osmet/file.rs:
```diff
@@ -0,0 +1,132 @@
+pub(super) struct OsmetFile {
+    magic: [u8; 8],
+    major: u16,
+    minor: u16,
```
Actually, the data structures don't have any extension points, so feature flags don't make sense either.  Maybe just a single version number?

But also, if I add a field to `struct Osmet` and try to read an existing osmet file, I get:

```
Error: failed to deserialize osmet file
Caused by: io error: failed to fill whole buffer
```

because bincode [doesn't handle changes to struct definitions](http://tyoverby.com/posts/bincode_release.html#should-i-use-bincode).  We could have a separate, fixed header that includes just the magic number and version information (more limited than the one contemplated by https://github.com/coreos/coreos-installer/pull/187#discussion_r402690732).  Or we could use something other than bincode.

jlebon replied:
Yup, separating magic number + version info into a fixed header makes sense to me.

As mentioned in https://github.com/coreos/coreos-installer/pull/187#issuecomment-607445367, backwards compatibility isn't a concern here. IOW, maybe it's not even worth a "version" field per se.

How about just a hash of the coreos-installer [version field](https://github.com/coreos/coreos-installer/blob/893f87820d09eca0a92ecf04c7f02536f930d76f/Cargo.toml#L9) itself? The main thing I want to enforce is making sure that the same code that did the packing is also doing the unpacking and error out otherwise.

bgilbert replied:
Could do, though that won't catch semantic changes among development builds within a release.  I'd be inclined to stick to a single version field (defined in a symbolic constant) and hand-bump it on changes.

jlebon commented on src/osmet/io_helpers.rs:
```diff
@@ -0,0 +1,191 @@
+    if bytes_copied != n {
+        bail!("expected to copy {} but instead copied {}", n, bytes_copied);
+    }
+    Ok(n)
```
Yeah, though it's helpful to match the same semantics so we can keep using e.g. `counter += copy_exactly_n(...)`

jlebon commented on src/osmet/io_helpers.rs:
```diff
@@ -0,0 +1,191 @@
+        .unwrap();
+    assert_eq!(chksum2.len(), 2);
+    let chksum62 = path.file_stem().unwrap().to_str().unwrap();
+    assert_eq!(chksum62.len(), 62);
```
Yeah, makes sense. (I mean, if the path isn't as expected, then there's an odd-looking file in the OSTree repo object store, which is super dubious. But it's silly to just throw that out instead of printing something out that will help us debug.)

jlebon commented on src/osmet/fiemap.rs:
```diff
@@ -0,0 +1,173 @@
+            } else if extent.fe_flags & ffi::FIEMAP_EXTENT_DELALLOC > 0 {
+                return Err("extent not allocated yet".into());
+            } else if extent.fe_flags & ffi::FIEMAP_EXTENT_UNWRITTEN > 0 {
+                return Err("extent not synced yet".into());
```
Do you have an alternative error string?

bgilbert replied:
`extent preallocated`?

jlebon commented on src/osmet/file.rs:
```diff
@@ -0,0 +1,132 @@
+            );
+        }
+        verify_canonical(&partition.mappings)?;
+        cursor = partition.end_offset;
```
Great catch!

jlebon commented on src/osmet/fiemap.rs:
```diff
@@ -0,0 +1,173 @@
+            if extent.fe_flags & ffi::FIEMAP_EXTENT_NOT_ALIGNED > 0 {
+                return Err("extent not aligned".into());
+            } else if extent.fe_flags & ffi::FIEMAP_EXTENT_MERGED > 0 {
+                return Err("file does not support extents".into());
```
It's not strictly an error, but nothing in CoreOS triggers this right now and I'd like to error out for now so that we notice it. Added a comment to clarify this.

jlebon commented on src/osmet/mod.rs:
```diff
@@ -0,0 +1,579 @@
+    // verify that re-packing will yield the expected checksum
+    println!("Verifying that repacked image matches digest");
+    let (checksum, unpunched_size) =
+        get_unpacked_image_digest(&mut punched_image, &partitions, &root)?;
```
I think that would make sense, yeah. Mind if we do that as a follow-up?

bgilbert replied:
Works for me.

jlebon commented on src/blockdev.rs:
```diff
@@ -224,13 +273,38 @@ pub fn get_sector_size(file: &File) -> Result<NonZeroU32> {
+/// Get the size of a block device.
+pub fn get_block_size(file: &File) -> Result<NonZeroU64> {
+    let fd = file.as_raw_fd();
+    let mut size: libc::size_t = 0;
```
Hmm, don't see this. Link?

bgilbert replied:
Whoops, nope, I was looking at bad non-canonical docs.

jlebon commented on src/osmet/mod.rs:
```diff
@@ -0,0 +1,579 @@
+                .chain_err(|| "while writing in between extents")?;
+        }
+
+        // this is the crucial space-saving step; we skip over the extent we have a mapping for
```
I've standardized on the following terms throughout the osmet codebase:
- the "unpacked" image refers to the fully blown up metal image
- extents for which we already have a mapping are "skipped" (instead of "punched")
- the "packed" image refers to the metal image with all the extents for which we already have a mapping skipped
- the "xzpacked" image is the packed image compressed with xz

bgilbert replied:
Could we get the above in a comment?

jlebon commented on src/osmet/mod.rs:
```diff
@@ -0,0 +1,579 @@
+
+    let mut mappings: Vec<Mapping> = vec![];
+
+    println!("Total /boot extents: {}", boot_files.len());
```
Good catch!

jlebon commented:
OK, updated!

One thing worth noting is that I've merged most of https://github.com/coreos/coreos-installer/pull/197 into this PR. As mentioned in https://github.com/coreos/coreos-installer/pull/197#discussion_r411498576, there is now an `--offline` switch which is useful on its own, but will also allow us to test this path before entirely defaulting to it.

cgwalters commented on src/osmet/unpacker.rs:
```diff
@@ -0,0 +1,271 @@
+    let repo = {
+        let mut root_path = root.mountpoint().to_path_buf();
+        root_path.push("ostree/repo");
+        root_path
```
Isn't this just `let repo = root.mountpoint().join("ostree/repo")`?

jlebon replied:
Thanks, tweaked!

jlebon commented:
CI failure here due to network flake, though it did lead me to #222.

cgwalters commented:
Anything else to do to land this?

cgwalters commented:
One thing I can probably help with is adding `kola testiso --scenario osmet` once the "use osmet by default" PR lands too - it'd just be omitting the `coreos.inst.image_url` arg and we could also entirely omit networking from the VM too right?

jlebon commented:
> Anything else to do to land this?

I was hoping for @bgilbert to have another look at this before merging since he did such an awesome review. But I think he's been away this week. I do think it's in a pretty good shape though and that we can fixup things as follow-ups. Maybe let's give it until Monday?

> One thing I can probably help with is adding `kola testiso --scenario osmet` once the "use osmet by default" PR lands too - it'd just be omitting the `coreos.inst.image_url` arg and we could also entirely omit networking from the VM too right?

Thanks, that'd be awesome!

Actually, with this PR, we could have a test already that uses `--offline` for now with networking turned off in the ISO embed scenario.

And thinking more on this... since it's way more work to respin coreos-installer than cosa, maybe what we should do is land https://github.com/coreos/coreos-assembler/pull/1244 *first*, but with the osmet packing step behind a switch to `buildextend-live`. And then we can test this PR in CI right here! Hmm, yeah let me do that.

jlebon commented:
One thing I wanted to note here is that I tested osmet in earnest on RHCOS yesterday (previously, I had only sanity-checked that `coreos-installer osmet fiemap /usr/bin/foo` didn't error out on `FIEMAP_EXTENT_ENCODED`).

And we can make it work, but the LUKS container complicates things. So the code here will need to be adapted to handle that (essentially adding to the mapping offsets the offset of the LUKS data).

IOW, this will not work yet as is on RHCOS.

cgwalters commented:
> And we can make it work, but the LUKS container complicates things.

Ug.  It probably wouldn't be seriously hard to rework the rhcos luks bits to just leave a bit of space for the luks header before the root partition, and when LUKS is enabled, just move the partition back and generate the header only when it's required, that would allow us to drop all the LUKS bits from almost everywhere.  Including the dm-linear bit.

jlebon commented:
> Ug. It probably wouldn't be seriously hard to rework the rhcos luks bits to just leave a bit of space for the luks header before the root partition, and when LUKS is enabled, just move the partition back and generate the header only when it's required, that would allow us to drop all the LUKS bits from almost everywhere. Including the dm-linear bit.

That'd definitely make things simpler. Not that it's hard to handle it, but it feels awkward to do. From my experimentation yesterday, we have to `cryptsetup luksOpen` in osmet-pack in cosa for osmet to access the OSTree repo, and here, we have to special-case cipher_null LUKS containers and call out to `cryptsetup luksDump` to read the data offset.

Is the amount of work required to do the header fiddling work (and all the simplifications that ensue) justified given that (IIUC) the end goal is to get rid of it entirely in favor of https://github.com/coreos/fedora-coreos-tracker/issues/94?

jlebon commented:
Let's hold this until tomorrow morning. @bgilbert told me he should be able to take another look then. (And actually, I still want to get the cosa bits in so we can test it in CI here.)

bgilbert commented on src/osmet/io_helpers.rs:
```diff
@@ -0,0 +1,185 @@
+        .unwrap()
+        .to_str()
+        .unwrap();
+    let chksum62 = path.file_stem().unwrap().to_str().unwrap();
```
The other `unwrap()`s are arguably justifiable (though error returns would be better) but the last one will assert if there's a non-UTF-8 filename in the object directory.

bgilbert commented on src/osmet/unpacker.rs:
```diff
@@ -0,0 +1,266 @@
+    }
+
+    pub fn new_from_sysroot(osmet: &Path) -> Result<Self> {
+        let (_, osmet, punched_image) = osmet_file_read(&osmet)?;
```
s/punched/packed/

bgilbert commented on src/source.rs:
```diff
@@ -286,6 +303,59 @@ impl ImageLocation for StreamLocation {
+            reader: Box::new(unpacker),
+            length_hint: Some(length),
+            signature: None,
+            filename: "image.raw".to_string(),
```
It'd be "more correct" to base this on the osmet filename, but it doesn't affect any live codepaths right now.  :man_shrugging:

jlebon commented:
OK, addressed comments and added a second commit here which turns on CI testing via osmet! Feel free to do another final review on the latest minor changes, though otherwise will merge on green!

bgilbert commented on src/osmet/mod.rs:
```diff
@@ -0,0 +1,680 @@
+// limitations under the License.
+
+// Note the following terms are in use in this module:
+// - the "unpacked" image refers to the fully blown up metal image (as it'd be read from a blockd
```
`block`

bgilbert commented on src/source.rs:
```diff
@@ -286,6 +303,72 @@ impl ImageLocation for StreamLocation {
+        let unpacker = OsmetUnpacker::new_from_sysroot(Path::new(&self.osmet_path))?;
+
+        let filename = {
+            let mut filename = self.osmet_path.clone();
```
This is the entire path, right?  The field should have just the filename.

jlebon replied:
Man, I should've just stuck with `image.raw` for this... ;)  Anyway, I went the whole way with this now and it somehow spans 18 lines post-`cargo fmt`. :tada: 

(Though I do think in the future it'd be nice to have `download` be able to get back the raw image without actually writing it to a device, so this would be useful then.)

bgilbert replied:
Yeah... :disappointed:  Thanks for doing this.

jlebon commented:
> Successfully tested scenario iso-offline-install for 31.20200430.dev.0 on bios (metal)

:tada:

dustymabe commented:
Yay. this works! My cosa run made a metal image that looks like:


```
$ wc --bytes builds/latest/x86_64/fedora-coreos-31.20200430.dev.0-metal.x86_64.raw
2959081472 builds/latest/x86_64/fedora-coreos-31.20200430.dev.0-metal.x86_64.raw

$ sha256sum builds/latest/x86_64/fedora-coreos-31.20200430.dev.0-metal.x86_64.raw
9dd23ee016f4c9e8562661dc6cce92f90a43e727f9254f2f671383aecc87a413  builds/latest/x86_64/fedora-coreos-31.20200430.dev.0-metal.x86_64.raw
```

I booted to the Live ISO and run the installer in offline mode!

```
$ sudo coreos-installer install /dev/sda --offline
Installing Fedora CoreOS 31.20200430.dev.0 x86_64 (512-byte sectors)
> Read disk 2.8 GiB/2.8 GiB (100%)
Install complete.
```

And then I read back the written out contents and it matches the same
checksum of the metal image!

```
$ sudo dd if=/dev/sda bs=512 count=5779456 | sha256sum
5779456+0 records in
5779456+0 records out
2959081472 bytes (3.0 GB, 2.8 GiB) copied, 24.9252 s, 119 MB/s
9dd23ee016f4c9e8562661dc6cce92f90a43e727f9254f2f671383aecc87a413  -
```

Nice work @jlebon!
